### DISCUSSION TITLE
Webview widget / custom widgets in rust

### DISCUSSION THREAD

**Discussion by MichaelAkvo** [slint-ui/slint#1489] [Q&A]
Hi,

My colleagues wanted to use JS libraries to draw graphs and maps in their application. That's just to use existing libraries and not reinvent the wheel. Unfortunately, there's no webview widget in slint and I can't find an easy way to add it. I was looking to see if it's possible to write custom widgets, but the closest I could find for this the opengl_underlay, however that doesn't seem to get a context for the widget, but the entire window.

To be fair, I haven't exactly looked at what webkit or blink needs in order to be embedded into an application.

Before I or somebody else gets started on this, is a webview widget planned? If not, how would one write a new widget that can be used in `.slint`? I'm not sure if there's a DOM to just insert new widgets in different places or if the widget class has to be registered somewhere in order to be accessible to `.slint`

**Comment by ogoffart** [comment]
A web view is definitively something we plan to do, but not something we will do in the short term

There could be several ways to implement it.
One would first need to choose a library such as blink or webkit to do the actual web rendering.
Then we could either use the GL api to have texture drawn,  or first render to an Image and render that.  And forward the mouse event as well.

If done within slint using its private api from i-slint-core,  one would have to create an Item, the same way that we render the native style such as  `NativeButton` and such.

Otherwise, this could be done in .slint using an Image a bit like the plotter example: https://slint-ui.com/snapshots/master/demos/plotter/

**Reply by MichaelAkvo** [reply]
Interesting. I'll have a look at the plotter first. Using prior art would probably be simpler than trying to figure out the private API. It will presumably be orders of a magnitude slower than providing a GL context but a start which could probably be reused.

Is there a public API in planning to add custom `Item`s?

**Comment by raphael10-collab** [comment]
@ogoffart What about the webview library? https://github.com/webview/webview . How would you suggest to proceed in order to use this tiny library with Slint ?

**Reply by tronical** [reply]
The web view library looks promising. I see two issues:

1. The library requires window-level embedding, which we don't support in Slint. We could try to implement this in the long term, with all the limitations it brings.
2. On Linux it requires the parent window to be a Gtk+ window, which won't work with winit nor Qt. This is not even web view specific, also Tauri's web view wrappers use WebKitGtk on Linux. I don't see an easy solution here. On Linux WPE would probably provide better graphical primitives for an integration.

**Reply by raphael10-collab** [reply]
@tronical  Interesting analysis. How to embed a web page (for example https://slint.dev/) in the slint window, then? Any steps forward on the track traced by @ogoffart https://github.com/slint-ui/slint/discussions/1489#discussioncomment-3430681 ?
